Makefile

```diff
@@ -34,8 +34,10 @@ UNAME := $(shell uname)
 .SILENT:
 
 all: libs
+ifndef FREEBSD

```

Comment:
We need to find a solution for building and running the unit-test projects...

Reply:
This is beyond my pay grade. I will leave it to you. I leave my instruction to how to grab a FreeBSD env. TBH, I'm not expert on FreeBSD. I learnt FreeBSD because of patching a port for my pfsense router. See [my wiki here](https://github.com/rickyzhang82/FreeBSDWiki/wiki/1.-Setup-Development-Environment-for-pfSense-FreeBSD-port) to guide you.

Reply:
I will also try. Do you see a build issue or a runtime error?

Reply:
No, the build is clean and one of the app runs fine.

Reply:
I'd guess that probably `Packet++Test` runs fine and `Pcap++Test` fails? Or is it the other way around?

Reply:
It failed in linking mem leak check library with two tests.

Reply:
I can see `memplumber.o` compiled in `PcapPlusPlus/3rdParty/MemPlumber/MemPlumber`. 

But it failed to link with memplumber.o. Both `backtrace` and `backtrace_symbols` are declared in FreeBSD header `execinfo.h`. I did found the header `execinfo.h` in my file system.

I'm scratching my head what went wrong.

```
==> Building target: Packet++Test
Building file: main.cpp
In file included from main.cpp:32:
In file included from ../PcppTestFramework/PcppTestFramework.h:4:
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.h:14:6: warning: function previously declared with an explicit exception specification redeclared with an implicit exception specification [-Wimplicit-exception-spec-mismatch]
void operator delete[](void* pointer);  
     ^
/usr/include/c++/v1/new:191:36: note: previous declaration is here
_LIBCPP_OVERRIDABLE_FUNC_VIS void  operator delete[](void* __p) _NOEXCEPT;
                                   ^
In file included from main.cpp:32:
In file included from ../PcppTestFramework/PcppTestFramework.h:4:
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.h:15:6: warning: function previously declared with an explicit exception specification redeclared with an implicit exception specification [-Wimplicit-exception-spec-mismatch]
void operator delete(void* pointer, std::size_t size);
     ^
/usr/include/c++/v1/new:186:74: note: previous declaration is here
_LIBCPP_OVERRIDABLE_FUNC_VIS _LIBCPP_AVAILABILITY_SIZED_NEW_DELETE void  operator delete(void* __p, std::size_t __sz) _NOEXCEPT;
                                                                         ^
In file included from main.cpp:32:
In file included from ../PcppTestFramework/PcppTestFramework.h:4:
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.h:16:6: warning: function previously declared with an explicit exception specification redeclared with an implicit exception specification [-Wimplicit-exception-spec-mismatch]
void operator delete[](void* pointer, std::size_t size);
     ^
/usr/include/c++/v1/new:194:74: note: previous declaration is here
_LIBCPP_OVERRIDABLE_FUNC_VIS _LIBCPP_AVAILABILITY_SIZED_NEW_DELETE void  operator delete[](void* __p, std::size_t __sz) _NOEXCEPT;
                                                                         ^
3 warnings generated.
Building MemPlumber file: /home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:405:6: warning: function previously declared with an explicit exception specification
      redeclared with an implicit exception specification [-Wimplicit-exception-spec-mismatch]
void operator delete(void* pointer, std::size_t size) {
     ^
/usr/include/c++/v1/new:186:74: note: previous declaration is here
_LIBCPP_OVERRIDABLE_FUNC_VIS _LIBCPP_AVAILABILITY_SIZED_NEW_DELETE void  operator delete(void* __p, std::size_t __sz) _NOEXCEPT;
                                                                         ^
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:409:6: warning: function previously declared with an explicit exception specification
      redeclared with an implicit exception specification [-Wimplicit-exception-spec-mismatch]
void operator delete[](void* pointer) {
     ^
/usr/include/c++/v1/new:191:36: note: previous declaration is here
_LIBCPP_OVERRIDABLE_FUNC_VIS void  operator delete[](void* __p) _NOEXCEPT;
                                   ^
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:413:6: warning: function previously declared with an explicit exception specification
      redeclared with an implicit exception specification [-Wimplicit-exception-spec-mismatch]
void operator delete[](void* pointer, std::size_t size) {
     ^
/usr/include/c++/v1/new:194:74: note: previous declaration is here
_LIBCPP_OVERRIDABLE_FUNC_VIS _LIBCPP_AVAILABILITY_SIZED_NEW_DELETE void  operator delete[](void* __p, std::size_t __sz) _NOEXCEPT;
                                                                         ^
3 warnings generated.
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.o: In function `getCaller()':
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:359: undefined reference to `backtrace'
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:366: undefined reference to `backtrace_symbols'
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:359: undefined reference to `backtrace'
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:366: undefined reference to `backtrace_symbols'
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:359: undefined reference to `backtrace'
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:366: undefined reference to `backtrace_symbols'
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:359: undefined reference to `backtrace'
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:366: undefined reference to `backtrace_symbols'
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:359: undefined reference to `backtrace'
/home/Ricky/repo/github/PcapPlusPlus/3rdParty/MemPlumber/MemPlumber/memplumber.cpp:366: undefined reference to `backtrace_symbols'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[1]: *** [Makefile:54: Packet++Test] Error 1
gmake: *** [Makefile:37: all] Error 2

```

Reply:
I think I found the culprit. You haven't explicitly link `-lexecinfo` for libexecinfo.

Reply:
Check out my commit 0bb75edd. I enable test compilation. All tests get compiled now.

How do you trigger the tests to run by make?

Reply:
this is great news!!! thanks for looking into it!!!

it is currently not possible to trigger the tests from `make`. What you should is finish building PcapPlusPlus and then run the following:

```
cd <PCAPPLUSPLUS_LOCATION>/Tests/Packet++Test
Bin/Packet++Test

cd <PCAPPLUSPLUS_LOCATION>/Tests/Pcap++Test
sudo Bin/Pcap++Test -i <IP_ADDRESS>
```

Reply:
Below are the test result.

Packet++ test are all passed.
```
Ricky@gtx-freebsd ~/repo/github/PcapPlusPlus/Tests/Packet++Test $ ./Bin/Packet++Test 
PcapPlusPlus version: v19.04+ (non-official release)
Built: Oct 23 2019 16:54:03
Built from: Git branch 'dev-pr-freebsd-build', commit '92b53d12dc1d645816478f3818f393a6e228d097'
Start running tests...

EthPacketCreation             : PASSED
EthPacketPointerCreation      : PASSED
EthAndArpPacketParsing        : PASSED
ArpPacketCreation             : PASSED
VlanParseAndCreation          : PASSED
Ipv4PacketCreation            : PASSED
Ipv4PacketParsing             : PASSED
Ipv4FragmentationTest         : PASSED
Ipv4OptionsParsingTest        : PASSED
Ipv4OptionsEditTest           : PASSED
Ipv4UdpChecksum               : PASSED
Ipv6UdpPacketParseAndCreate   : PASSED
Ipv6ExtensionsTest            : PASSED
TcpPacketNoOptionsParsing     : PASSED
TcpPacketWithOptionsParsing   : PASSED
TcpPacketWithOptionsParsing2  : PASSED
TcpPacketCreation             : PASSED
TcpPacketCreation2            : PASSED
InsertDataToPacket            : PASSED
InsertVlanToPacket            : PASSED
RemoveLayerTest               : PASSED
HttpRequestLayerParsingTest   : PASSED
HttpRequestLayerCreationTest  : PASSED
HttpRequestLayerEditTest      : PASSED
HttpResponseLayerParsingTest  : PASSED
HttpResponseLayerCreationTest : PASSED
HttpResponseLayerEditTest     : PASSED
PPPoESessionLayerParsingTest  : PASSED
PPPoESessionLayerCreationTest : PASSED
PPPoEDiscoveryLayerParsingTest: PASSED
PPPoEDiscoveryLayerCreateTest : PASSED
DnsLayerParsingTest           : PASSED
DnsLayerQueryCreationTest     : PASSED
DnsLayerResourceCreationTest  : PASSED
DnsLayerEditTest              : PASSED
DnsLayerRemoveResourceTest    : PASSED
MplsLayerTest                 : PASSED
CopyLayerAndPacketTest        : PASSED
IcmpParsingTest               : PASSED
IcmpCreationTest              : PASSED
IcmpEditTest                  : PASSED
GreParsingTest                : PASSED
GreCreationTest               : PASSED
GreEditTest                   : PASSED
SSLClientHelloParsingTest     : PASSED
SSLAppDataParsingTest         : PASSED
SSLAlertParsingTest           : PASSED
SSLMultipleRecordParsingTest  : PASSED
SSLMultipleRecordParsing2Test : PASSED
SSLMultipleRecordParsing3Test : PASSED
SSLMultipleRecordParsing4Test : PASSED
SSLPartialCertificateParseTest: PASSED
SSLNewSessionTicketParseTest  : PASSED
SllPacketParsingTest          : PASSED
SllPacketCreationTest         : PASSED
DhcpParsingTest               : PASSED
DhcpCreationTest              : PASSED
DhcpEditTest                  : PASSED
NullLoopbackTest              : PASSED
IgmpParsingTest               : PASSED
IgmpCreateAndEditTest         : PASSED
Igmpv3ParsingTest             : PASSED
Igmpv3QueryCreateAndEditTest  : PASSED
Igmpv3ReportCreateAndEditTest : PASSED
ParsePartialPacketTest        : PASSED
VxlanParsingAndCreationTest   : PASSED
SipRequestLayerParsingTest    : PASSED
SipRequestLayerCreationTest   : PASSED
SipRequestLayerEditTest       : PASSED
SipResponseLayerParsingTest   : PASSED
SipResponseLayerCreationTest  : PASSED
SipResponseLayerEditTest      : PASSED
SdpLayerParsingTest           : PASSED
SdpLayerCreationTest          : PASSED
SdpLayerEditTest              : PASSED
PacketTrailerTest             : PASSED
RadiusLayerParsingTest        : PASSED
RadiusLayerCreationTest       : PASSED
RadiusLayerEditTest           : PASSED
GtpLayerParsingTest           : PASSED
GtpLayerCreationTest          : PASSED
GtpLayerEditTest              : PASSED
ALL TESTS PASSED!!
```

Some Pcap++ failed. Note that I'm running FreeBSD inside my VirtualBox. The NIC is a bridged adapter. Maybe you can shed some light why it failed.

```
Ricky@gtx-freebsd ~/repo/github/PcapPlusPlus/Tests/Pcap++Test $ ifconfig
em0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
	options=9b<RXCSUM,TXCSUM,VLAN_MTU,VLAN_HWTAGGING,VLAN_HWCSUM>
	ether 08:00:27:bf:b2:a5
	hwaddr 08:00:27:bf:b2:a5
	inet6 fe80::a00:27ff:febf:b2a5%em0 prefixlen 64 scopeid 0x1 
	inet 192.168.2.29 netmask 0xfffffc00 broadcast 192.168.3.255 
	nd6 options=23<PERFORMNUD,ACCEPT_RTADV,AUTO_LINKLOCAL>
	media: Ethernet autoselect (1000baseT <full-duplex>)
	status: active
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> metric 0 mtu 16384
	options=600003<RXCSUM,TXCSUM,RXCSUM_IPV6,TXCSUM_IPV6>
	inet6 ::1 prefixlen 128 
	inet6 fe80::1%lo0 prefixlen 64 scopeid 0x2 
	inet 127.0.0.1 netmask 0xff000000 
	nd6 options=23<PERFORMNUD,ACCEPT_RTADV,AUTO_LINKLOCAL>
	groups: lo 

Ricky@gtx-freebsd ~/repo/github/PcapPlusPlus/Tests/Pcap++Test $ sudo ./Bin/Pcap++Test -i 192.168.2.29 -s
Skipping memory leak check for all test cases
PcapPlusPlus version: v19.04+ (non-official release)
Built: Oct 23 2019 16:54:10
Git info: Git branch 'dev-pr-freebsd-build', commit '92b53d12dc1d645816478f3818f393a6e228d097'
Using ip: 192.168.2.29
Debug mode: off
Starting tests...
Start running tests...

TestIPAddress                 : PASSED
TestMacAddress                : PASSED
TestPcapFileReadWrite         : PASSED
TestPcapSllFileReadWrite      : PASSED
TestPcapRawIPFileReadWrite    : PASSED
TestPcapFileAppend            : PASSED
TestPcapNgFileReadWrite       : PASSED
TestPcapNgFileReadWriteAdv    : PASSED
TestPcapLiveDeviceList        : FAILED. assertion failed: Couldn't find default gateway for any of the interfaces
TestPcapLiveDeviceListSearch  : PASSED
TestPcapLiveDevice            : PASSED
TestPcapLiveDeviceNoNetworking: FAILED. assertion failed: Cannot find MAC address for device 'em0'
TestPcapLiveDeviceStatsMode   : PASSED
TestPcapLiveDeviceBlockingMode: PASSED
TestPcapLiveDeviceSpecialCfg  : PASSED
TestWinPcapLiveDevice         : PASSED
TestPcapLiveDeviceByInvalidIp : PASSED
TestPcapFiltersLive           : PASSED
TestPcapFilters_General_BPFStr: PASSED
TestPcapFiltersOffline        : PASSED
Packet length [1501] is larger than device MTU [1500]

TestSendPacket                : PASSED
TestSendPackets               : PASSED
TestRemoteCapture             : PASSED
TestHttpRequestParsing        : PASSED
TestHttpResponseParsing       : PASSED
TestPrintPacketAndLayers      : PASSED
TestPfRingDevice              : SKIPPED (PF_RING not configured)
TestPfRingDeviceSingleChannel : SKIPPED (PF_RING not configured)
TestPfRingMultiThreadAllCores : SKIPPED (PF_RING not configured)
TestPfRingMultiThreadSomeCores: SKIPPED (PF_RING not configured)
TestPfRingSendPacket          : SKIPPED (PF_RING not configured)
TestPfRingSendPackets         : SKIPPED (PF_RING not configured)
TestPfRingFilters             : SKIPPED (PF_RING not configured)
TestDnsParsing                : PASSED
TestDpdkDevice                : SKIPPED (DPDK not configured)
TestDpdkMultiThread           : SKIPPED (DPDK not configured)
TestDpdkDeviceSendPackets     : SKIPPED (DPDK not configured)
TestKniDevice                 : SKIPPED (DPDK not configured)
TestKniDeviceSendReceive      : SKIPPED (DPDK not configured)
TestDpdkMbufRawPacket         : SKIPPED (DPDK not configured)
TestDpdkDeviceWorkerThreads   : SKIPPED (DPDK not configured)
TestGetMacAddress             : FAILED. assertion failed: Arping to all IPs in arp-table failed
TestTcpReassemblySanity       : PASSED
TestTcpReassemblyRetran       : PASSED
TestTcpReassemblyMissingData  : PASSED
TestTcpReassemblyOutOfOrder   : PASSED
TestTcpReassemblyWithFIN_RST  : PASSED
TestTcpReassemblyMalformedPkts: PASSED
TestTcpReassemblyMultipleConns: PASSED
TestTcpReassemblyIPv6         : PASSED
TestTcpReassemblyIPv6MultConns: PASSED
TestTcpReassemblyIPv6_OOO     : PASSED
TestTcpReassemblyCleanup      : PASSED
TestIPFragmentationSanity     : PASSED
TestIPFragOutOfOrder          : PASSED
TestIPFragPartialData         : PASSED
TestIPFragMultipleFrags       : PASSED
TestIPFragMapOverflow         : PASSED
TestIPFragRemove              : PASSED
Raw socket are not supported on this platform
TestRawSockets                : FAILED. assertion failed: Couldn't open raw socket
NOT ALL TESTS PASSED!!
```

Reply:
Regarding to the live capture devices enumeration, I tested `TcpReassembly` in my pfsense router. I can list all NIC in my bare metals correctly.

Reply:
Thanks for posting the unit-test results!

Regarding the failed tests:

- `TestPcapLiveDeviceList` - setting the default gateway is done differently on every OS. Here is the code that does it:  https://github.com/seladb/PcapPlusPlus/blob/f25cf80f5fe4c5f9b837edfda775c57e04585d36/Pcap%2B%2B/src/PcapLiveDevice.cpp#L775. We should probably add a section for FreeBSD
- `TestPcapLiveDeviceNoNetworking` - setting MacAddress is done differently on every OS. Here is the code that does it: https://github.com/seladb/PcapPlusPlus/blob/f25cf80f5fe4c5f9b837edfda775c57e04585d36/Pcap%2B%2B/src/PcapLiveDevice.cpp#L688. We should probably add a section for FreeBSD
- `TestGetMacAddress` - I think this test fails because it can't find the MacAddress for the interface, which is the same issue as we have in `TestPcapLiveDeviceList`
- `TestRawSockets` - I'm not sure if raw sockets are supported on FreeBSD or not and what level of support is there. We should probably either fix the code or fix the tests

Reply:
Sorry, I can't figure out how your code works in `PcapLiveDevice.cpp`. The `void PcapLiveDevice::setDeviceMacAddress()` seem to get Mac address rather than set another one. This is really confusing.

In any case, I blindly applied a patch below:

```
Ricky@gtx-freebsd ~/repo/github/PcapPlusPlus/Tests/Pcap++Test $ git diff
diff --git a/Pcap++/src/PcapLiveDevice.cpp b/Pcap++/src/PcapLiveDevice.cpp
index dd73a43..d0d8540 100644
--- a/Pcap++/src/PcapLiveDevice.cpp
+++ b/Pcap++/src/PcapLiveDevice.cpp
@@ -25,7 +25,7 @@
 #include <sys/sysctl.h>
 #include <net/if.h>
 #endif
-#ifdef MAC_OS_X
+#if defined(MAC_OS_X) || defined(FREEBSD)^M
 #include <net/if_dl.h>
 #endif
 
@@ -737,7 +737,7 @@ void PcapLiveDevice::setDeviceMacAddress()
        }
 
     m_MacAddress = MacAddress(ifr.ifr_hwaddr.sa_data[0], ifr.ifr_hwaddr.sa_data[1], ifr.ifr_hwaddr.sa_data[2], ifr.ifr_hwaddr.sa_data[3], ifr.ifr_hwaddr.sa_data[4], ifr.ifr_hwaddr.sa_data[5]);
-#elif MAC_OS_X
+#elif MAC_OS_X || FREEBSD^M
     int        mib[6];
     size_t len;
 
@@ -831,7 +831,7 @@ void PcapLiveDevice::setDefaultGateway()
            interfaceGatewayStream >> interfaceGatewayIPInt;
            m_DefaultGateway = IPv4Address(interfaceGatewayIPInt);
        }
-#elif MAC_OS_X
+#elif MAC_OS_X || FREEBSD^M
        std::string ifaceStr = std::string(m_Name);
        std::string command = "netstat -nr | grep default | grep " + ifaceStr;
        std::string ifaceInfo = executeShellCommand(command);
```
Then, run the test. It passed except raw socket.

```
Ricky@gtx-freebsd ~/repo/github/PcapPlusPlus/Tests/Pcap++Test $ sudo ./Bin/Pcap++Test -i 192.168.2.29
PcapPlusPlus version: v19.04+ (non-official release)
Built: Oct 23 2019 16:54:10
Git info: Git branch 'dev-pr-freebsd-build', commit '92b53d12dc1d645816478f3818f393a6e228d097'
Using ip: 192.168.2.29
Debug mode: off
Starting tests...
Start running tests...

Error in retrieving MAC address: if_nametoindex error
Error in retrieving MAC address: if_nametoindex error
TestIPAddress                 : PASSED
TestMacAddress                : PASSED
TestPcapFileReadWrite         : PASSED
TestPcapSllFileReadWrite      : PASSED
TestPcapRawIPFileReadWrite    : PASSED
TestPcapFileAppend            : PASSED
TestPcapNgFileReadWrite       : PASSED
TestPcapNgFileReadWriteAdv    : PASSED
Error in retrieving MAC address: if_nametoindex error
Error in retrieving MAC address: if_nametoindex error
TestPcapLiveDeviceList        : PASSED
TestPcapLiveDeviceListSearch  : PASSED
TestPcapLiveDevice            : PASSED
TestPcapLiveDeviceNoNetworking: PASSED
TestPcapLiveDeviceStatsMode   : PASSED
TestPcapLiveDeviceBlockingMode: PASSED
TestPcapLiveDeviceSpecialCfg  : PASSED
TestWinPcapLiveDevice         : PASSED
TestPcapLiveDeviceByInvalidIp : PASSED
TestPcapFiltersLive           : PASSED
TestPcapFilters_General_BPFStr: PASSED
TestPcapFiltersOffline        : PASSED
Packet length [1501] is larger than device MTU [1500]

TestSendPacket                : PASSED
TestSendPackets               : PASSED
TestRemoteCapture             : PASSED
TestHttpRequestParsing        : PASSED
TestHttpResponseParsing       : PASSED
TestPrintPacketAndLayers      : PASSED
TestPfRingDevice              : SKIPPED (PF_RING not configured)
TestPfRingDeviceSingleChannel : SKIPPED (PF_RING not configured)
TestPfRingMultiThreadAllCores : SKIPPED (PF_RING not configured)
TestPfRingMultiThreadSomeCores: SKIPPED (PF_RING not configured)
TestPfRingSendPacket          : SKIPPED (PF_RING not configured)
TestPfRingSendPackets         : SKIPPED (PF_RING not configured)
TestPfRingFilters             : SKIPPED (PF_RING not configured)
TestDnsParsing                : PASSED
TestDpdkDevice                : SKIPPED (DPDK not configured)
TestDpdkMultiThread           : SKIPPED (DPDK not configured)
TestDpdkDeviceSendPackets     : SKIPPED (DPDK not configured)
TestKniDevice                 : SKIPPED (DPDK not configured)
TestKniDeviceSendReceive      : SKIPPED (DPDK not configured)
TestDpdkMbufRawPacket         : SKIPPED (DPDK not configured)
TestDpdkDeviceWorkerThreads   : SKIPPED (DPDK not configured)
TestGetMacAddress             : PASSED
TestTcpReassemblySanity       : PASSED
TestTcpReassemblyRetran       : PASSED
TestTcpReassemblyMissingData  : PASSED
TestTcpReassemblyOutOfOrder   : PASSED
TestTcpReassemblyWithFIN_RST  : PASSED
TestTcpReassemblyMalformedPkts: PASSED
TestTcpReassemblyMultipleConns: PASSED
TestTcpReassemblyIPv6         : PASSED
TestTcpReassemblyIPv6MultConns: PASSED
TestTcpReassemblyIPv6_OOO     : PASSED
TestTcpReassemblyCleanup      : PASSED
TestIPFragmentationSanity     : PASSED
TestIPFragOutOfOrder          : PASSED
TestIPFragPartialData         : PASSED
TestIPFragMultipleFrags       : PASSED
TestIPFragMapOverflow         : PASSED
TestIPFragRemove              : PASSED
Raw socket are not supported on this platform
TestRawSockets                : FAILED. assertion failed: Couldn't open raw socket
NOT ALL TESTS PASSED!!
```

I have no better ideas how to address it. How about you take it over from here?

Reply:
The print message "Error in retrieving MAC address: if_nametoindex error" indicate that `void PcapLiveDevice::setDeviceMacAddress()` failed but some how previous failed test like `TestPcapLiveDeviceList`, `TestPcapLiveDeviceNoNetworking` and `TestGetMacAddress` pass with this patch. Any ideas?

Reply:
I guess the error message "Error in retrieving MAC address: if_nametoindex error" comes from iteration of the list of PcapLiveDevice. When it doesn't fail, it returns MAC address. That may explain why.

I don't know anything about raw socket in *nix system. 

If you are OK with my patch, I will include it in my PR. Please let me know.

Reply:
Here are the result of listing interface:
```
Ricky@gtx-freebsd ~/repo/github/PcapPlusPlus/Dist/examples $ sudo ./TcpReassembly -l
Error in retrieving MAC address: if_nametoindex error
Error in retrieving MAC address: if_nametoindex error

Network interfaces:
    -> Name: 'em0'   IP address: 192.168.2.29
    -> Name: 'lo0'   IP address: 127.0.0.1
    -> Name: 'usbus0'   IP address: 0.0.0.0
    -> Name: 'usbus1'   IP address: 0.0.0.0

```

Reply:
Regarding RawSocket - I don't think they're properly supported on FreeBSD and the code already assumes that. The reason the test fails is because of a bug I introduced a few months ago which you just uncovered. Please leave the code as is, I'll fix it later.

The reason for the `Error in retrieving MAC address: if_nametoindex error` is probably what you mentioned - getting the MAC address fails for the other interfaces. Just to make sure - can you please add a print `m_Name` when this error happens? If this is indeed the case we should probably turn this `LOG_ERROR()` into `LOG_DEBUG()`, unless you have a better idea how to solve this

Reply:
This is indeed the case:

```
Ricky@gtx-freebsd ~/repo/github/PcapPlusPlus/Tests/Pcap++Test $ sudo ./Bin/Pcap++Test -i 192.168.2.29
PcapPlusPlus version: v19.04+ (non-official release)
Built: Oct 24 2019 08:13:31
Git info: Git branch 'dev-pr-freebsd-build', commit '3d8f2894b2f7c4b4445b1b7489eb19cd457451fe'
Using ip: 192.168.2.29
Debug mode: off
Starting tests...
Start running tests...

m_Name: em0
m_Name: lo0
m_Name: usbus0
Error in retrieving MAC address: if_nametoindex error
m_Name: usbus1
Error in retrieving MAC address: if_nametoindex error
...
```

Can your downstream code handle the case that when `m_MacAddress` is an empty string? If yes, I think we should turn all three `LOG_ERROR()` into `LOG_DEBUG()` the code snippet below.

https://github.com/seladb/PcapPlusPlus/blob/f25cf80f5fe4c5f9b837edfda775c57e04585d36/Pcap%2B%2B/src/PcapLiveDevice.cpp#L751-L766

Reply:
I think the code already handles cases when `m_MacAddress` is an empty string, which means MAC address could not be retrieved. Did you find any place that doesn't?

Reply:
No. At least I don't see it crashed.

Done. see my commit 90f13bba

Reply:
Awesome thank you! So now all of the tests pass?

Reply:
No. The raw socket UT still failed in FreeBSD. Rest are clean in FreeBSD.

One of your test env failed in mysterious way by my commit 9a5df12, which introduce Cirrus CI. See [my comment](https://github.com/seladb/PcapPlusPlus/pull/261#issuecomment-546001746).

Reply:
The raw socket test failure is ok, I'll fix that later. I'm re-running this job to see if the issue persists

Reply:
I re-run the job and it's passing now. Not sure what the issue was...

Is there anything else to do before I can merge the PR?

Reply:
They all looks good now after you carefully review them.

You need to do:

- Fix raw socket
- Enable test in Cirrus yaml file. For FreeBSD, all tests in Packet++test are passed. Only one Pcap++test -- raw socket is failed. So you may skip raw socket for FreeBSD temporarily.

Reply:
Awesome! Let me merge your PR first and then add the fixes